File: src/hw3/hw3/data_sender.py
#!/usr/bin/env python3

# Example of a service in ROS 2
#
# data_sender.py
#
# Cory Hungate
#
# This is an example of a simple service client in ROS 2 that heavily draws on code 
# provided for the class by Bill Smart.


# standard imports.
import rclpy
from rclpy.node import Node
from rclpy.time import Time

from rob499_msgs.msg import TestPacket
from rob499_msgs.srv import SendData

# In my head, this is primarily a publisher node, hence the name
class BasicPublisher(Node):
	def __init__(self):
		super().__init__('service')

		# Creating the publisher and service
		self.pub = self.create_publisher(TestPacket, 'data', 10)
		self.service = self.create_service(SendData, 'send_data', self.send_data_service)
		
		#now some additional variables I'm gonna need
		self.timer = self.create_timer(2, self.callback)
		self.counter = 0
		self.send_data = True


	def callback(self):
		# Fill in the data in the response type.
		msg = TestPacket()
		msg.send_time = self.get_clock().now().to_msg()
		msg.payload = [1,2,3,4,5]		

		self.pub.publish(msg)

		# Log a message and increment the counter
		# self.get_logger().info(f'Got request # {self.counter} at time {msg.send_time}')
		self.counter += 1


	def send_data_service(self, request, response):
		self.send_data = request.send 

		if self.send_data:
			#if the data is stopped then restarted, we need to restart the timer
			if not self.timer or self.timer.is_canceled:
				self.timer = self.create_timer(2, self.callback)
			response.message = "Publishing started"
			self.get_logger().info("Started publishing data")
		else:
            # Stop the timer, effectively stopping the message publishing
			if self.timer:
				self.timer.cancel()
			response.message = "Publishing stopped"
			self.get_logger().info("Stopped publishing data")
		return response



# This is the entry point for the node.
def main(args=None):

	rclpy.init(args=args)

	publisher = BasicPublisher()

	rclpy.spin(publisher)
	rclpy.shutdown()


# This is the entry point when we call the node directly.
if __name__ == '__main__':
	main()

File: src/hw6/src/data_sender.cpp
// data_sender.cpp
//
// Cory Hungate
//
// A reimplementation of the data_sender node from HW3. This node publishes 
//a custom message called TestPacket on the topic 'data'. 


#include <rclcpp/rclcpp.hpp>

#include <rob499_msgs/msg/test_packet.hpp>
#include <rob499_msgs/srv/send_data.hpp>

#include <chrono>

using std::placeholders::_1;
using std::placeholders::_2;
using namespace std::chrono_literals;


// Service node that inherits from Node, as usual.
class SendDataService : public rclcpp::Node {
public:
	// Use an alias to simplify the syntax.
	using SendData = rob499_msgs::srv::SendData;
	using TestPacket = rob499_msgs::msg::TestPacket;

	SendDataService() :Node("send_data_service") {

		publisher_ = this->create_publisher<TestPacket>("data", 10);
		service_ = this->create_service<SendData>("send_data", std::bind(&SendDataService::service_callback, this, _1, _2));
		
		//if I have the default send_data_=true, then I want data to begin publishing immediately
		if (send_data_) {
			timer_ = this->create_wall_timer(0.5s, std::bind(&SendDataService::timer_callback, this));
		}
	}	

	

private:

	bool send_data_ = true;
	int counter_ = 0;

	rclcpp::Service<SendData>::SharedPtr service_;
	rclcpp::Publisher<TestPacket>::SharedPtr publisher_;
	rclcpp::TimerBase::SharedPtr timer_;

	// A callback to handle the service request.  The has parameters for the request and the response.
	void service_callback(
		const std::shared_ptr<SendData::Request> request,
		const std::shared_ptr<SendData::Response> response) {

		this->send_data_ = request->send;	

		if (this->send_data_ == true){
			if (this->timer_ == nullptr)
				this->timer_ = this->create_wall_timer(0.5s, std::bind(&SendDataService::timer_callback, this));
			response->message = "Data Publishing Started";
			RCLCPP_INFO(this->get_logger(), "Data is now publishing");
		}

		else{
			this->timer_.reset();
			response->message = "Data Publishing Stopped";
			RCLCPP_INFO(this->get_logger(), "Data is no longer publishing");
		}

	}

	void timer_callback(){
		auto msg = TestPacket();
		msg.send_time = this->now();
		msg.payload = {1,2,3,4,5}; 

		publisher_->publish(msg);

		//if you want to visualize the incoming data, just uncomment the next 2 lines
		// RCLCPP_INFO(this->get_logger(), "Receiver message #%i at time %.2f", 
		// 			this->counter_, rclcpp::Time(msg.send_time).seconds());

		this->counter_ += 1;

	}
};


int main(int argc, char **argv) {
	
	rclcpp::init(argc, argv);

	auto service = std::make_shared<SendDataService>();

	rclcpp::spin(service);

	rclcpp::shutdown();

	return 0;
}